include/envoy/network/connection.h

```diff
@@ -69,6 +69,51 @@ enum class ConnectionCloseType {
                      // until the delayed_close_timeout expires
 };
 
+/**
+ * Run-time modifiable connection options to alter socket settings
+ */
+class DynamicSocketOptions : public Logger::Loggable<Logger::Id::connection> {
```

Comment:
Thanks. This is moving in the right direction. I think we can improve on this by eliminating the need for the IoHandle parameter to the get and set methods of this class.
What you can do is: 
1.  make all of the get/set methods virtual and PURE
1.  make the Connection::getDynamicSocketOptionsPtr method PURE virtual as well.
1.  add DynamicSocketOptionsImpl : public DynamicSocketOptions class in the source/common/network/connection_impl.h
1.  The constructor to this class will accept the IoHandle& and store as its member. The implementation of the get/set methods would operate on the stored IoHandle
1.  Add ConnectionImpl::getDynamicSocketOptionsPtr implementation. You can make DynamicSocketOptionsImpl a member of the ConnectionImpl and initialize it in the constructor. The ConnectionImpl has the ioHandle() method to get the IoHandle.

This involves a bit more writing, but the user of the DynamicSocketOptions does not need to worry about the how to get the IoHandle. It may not always be possible if the used only has the Connection*.

Reply:
Thanks for the feedback. This makes sense, I will incorporate these changes. The other follow up question I had was is about the failure of my unit tests:

```
2019-12-17T09:25:18.5362942Z test/common/network/connection_impl_test.cc:799: Failure
2019-12-17T09:25:18.5363072Z Expected equality of these values:
2019-12-17T09:25:18.5363178Z   set_recv_lowat
2019-12-17T09:25:18.5363298Z     Which is: 4096
2019-12-17T09:25:18.5363418Z   get_recv_lowat
2019-12-17T09:25:18.5363519Z     Which is: 0
2019-12-17T09:25:18.5363731Z Stack trace:
2019-12-17T09:25:18.5363851Z   0x54785b: Envoy::Network::(anonymous namespace)::ConnectionImplTest_DynamicSocketOptions_Test::TestBody()
2019-12-17T09:25:18.5364236Z   0xab7c28: testing::internal::HandleExceptionsInMethodIfSupported<>()
2019-12-17T09:25:18.5364377Z   0xab7b35: testing::Test::Run()
2019-12-17T09:25:18.5364485Z   0xab89ef: testing::TestInfo::Run()
2019-12-17T09:25:18.5364615Z ... Google Test internal frames ...
2019-12-17T09:25:18.5364698Z 
2019-12-17T09:25:18.5365261Z [  FAILED  ] IpVersions/ConnectionImplTest.DynamicSocketOptions/IPv4, where GetParam() = 4-byte object <00-00 00-00> (9 ms)
2019-12-17T09:25:18.5365466Z [ RUN      ] IpVersions/ConnectionImplTest.DynamicSocketOptions/IPv6
2019-12-17T09:25:18.5365596Z test/common/network/connection_impl_test.cc:789: Failure
2019-12-17T09:25:18.5365727Z Expected equality of these values:
2019-12-17T09:25:18.5365854Z   get_recv_buf_size
2019-12-17T09:25:18.5365959Z     Which is: 0
2019-12-17T09:25:18.5366079Z   2 * set_recv_buf_size
2019-12-17T09:25:18.5366317Z     Which is: 8192
2019-12-17T09:25:18.5366440Z Stack trace:
2019-12-17T09:25:18.5366578Z   0x547738: Envoy::Network::(anonymous namespace)::ConnectionImplTest_DynamicSocketOptions_Test::TestBody()
2019-12-17T09:25:18.5366720Z   0xab7c28: testing::internal::HandleExceptionsInMethodIfSupported<>()
2019-12-17T09:25:18.5366856Z   0xab7b35: testing::Test::Run()
2019-12-17T09:25:18.5366980Z   0xab89ef: testing::TestInfo::Run()
2019-12-17T09:25:18.5367089Z ... Google Test internal frames ...
2019-12-17T09:25:18.5367184Z 
2019-12-17T09:25:18.5367292Z test/common/network/connection_impl_test.cc:799: Failure
2019-12-17T09:25:18.5404635Z Expected equality of these values:
2019-12-17T09:25:18.5404794Z   set_recv_lowat
2019-12-17T09:25:18.5404902Z     Which is: 4096
2019-12-17T09:25:18.5405014Z   get_recv_lowat
2019-12-17T09:25:18.5405123Z     Which is: 0
2019-12-17T09:25:18.5405223Z Stack trace:
2019-12-17T09:25:18.5405347Z   0x54785b: Envoy::Network::(anonymous namespace)::ConnectionImplTest_DynamicSocketOptions_Test::TestBody()
2019-12-17T09:25:18.5405506Z   0xab7c28: testing::internal::HandleExceptionsInMethodIfSupported<>()
2019-12-17T09:25:18.5405632Z   0xab7b35: testing::Test::Run()
2019-12-17T09:25:18.5405748Z   0xab89ef: testing::TestInfo::Run()
2019-12-17T09:25:18.5405856Z ... Google Test internal frames ...
2019-12-17T09:25:18.5405953Z 
2019-12-17T09:25:18.5406576Z [  FAILED  ] IpVersions/ConnectionImplTest.DynamicSocketOptions/IPv6, where GetParam() = 4-byte object <01-00 00-00> (1 ms)
```

Picked from https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/23705/logs/82

I do not see these failures on my local machine and the get and set values are reflected as expected. I run the envoy build container in privileged mode. Do you have any recommendations on how to fix these failures as well ?

Thanks for your time!

Reply:
Are you running build and test under docker?

Reply:
Yes, you are right. I am running the build and test under docker.

I pick the build image for from here:

`https://hub.docker.com/r/envoyproxy/envoy`

and spin up the container using this command:
```docker run  --privileged  -d -it --name envoy -v /home/rkavuluru/envoy:/envoy envoyproxy/envoy-build:latest```

Is this not the recommended approach ?

Reply:
It passes for me under the docker as well. It would be difficult to figure out why it fails without knowing the error code returned by the get/setsockopt.
My only suggestion at this point is to put printf() statements with the error codes, so we can see why it fails on the CI.

Reply:
Thanks for trying it out and the recommendation. I will get back with a patch for error codes and observe the CI pipeline.

Reply:
I was able to reproduce it after a bit. I have no explanation for the result though. getsockopt returns 0 value for the option and no error. I'm not sure what to suggest here. You can try asking on slack envoy-dev channel, to see if people may have suggestions.

Reply:
@yanavlasov thanks for the help. I tried it too and was able to reproduce this. It is not consistent; occurs after multiple restarts and goes away indefinitely. Also, I did some internal testing, what appears to me is that though read buffer values are 0; the application behaves as though the values for set buffers actually occurred.

I will post this on the dev channel; however I don't this it is as responsive. I will give it a shot.